devdoc/thandle_requirements.md

```diff
@@ -0,0 +1,135 @@
+`THANDLE_DEC_REF` decrements the reference count of `t` and frees memory if reference count reaches 0.
+
+**SRS_THANDLE_02_001: [** If `t` is `NULL` then `THANDLE_DEC_REF` shall return. **]**
+
+**SRS_THANDLE_02_002: [** `THANDLE_DEC_REF` shall decrement the ref count of `t`.  **]**
+
+**SRS_THANDLE_02_003: [** If the ref count of `t` reaches 0 then `THANDLE_DEC_REF` shall call `dispose` (if not `NULL`) and free the used memory.  **]**
+
+### THANDLE_INC_REF(T)
+```c
+MOCKABLE_FUNCTION(, void, THANDLE_INC_REF(T), THANDLE(T), t);
+```
+
+`THANDLE_INC_REF` increments the reference count of `t`.
+
+**SRS_THANDLE_02_004: [** If `t` is `NULL` then `THANDLE_INC_REF` shall return. **]**
+
+**SRS_THANDLE_02_005: [** `THANDLE_INC_REF` shall increment the reference count of `t`. **]**
+
+### THANDLE_ASSIGN(T)
+```c
+MOCKABLE_FUNCTION(, void, THANDLE_ASSIGN(T), THANDLE(T) *, t1, THANDLE(T), t2 );
+```
+
+`THANDLE_ASSIGN` does t1=t2. Both `t1` and `t2` are existing constructed handles.
+
+
+**SRS_THANDLE_02_006: [** If `t1` is `NULL` then `THANDLE_ASSIGN` shall return. **]**
+
+**SRS_THANDLE_02_007: [** If `*t1` is `NULL` and `t2` is `NULL` then `THANDLE_ASSIGN` shall return. **]**
+
+**SRS_THANDLE_02_008: [** If `*t1` is `NULL` and `t2` is not `NULL` then `THANDLE_ASSIGN` shall increment the reference count of `t2` and store `t2` in `*t1`. **]**
+
+**SRS_THANDLE_02_009: [** If `*t1` is not `NULL` and `t2` is `NULL` then `THANDLE_ASSIGN` shall decrement the reference count of `*t1` and store `NULL` in `*t1`. **]**
+
+**SRS_THANDLE_02_010: [** If `*t1` is not `NULL` and `t2` is not `NULL` then `THANDLE_ASSIGN` shall increment the reference count of `t2`, shall decrement the reference count of `*t1` and store `t2` in `*t1`. **]**
+
+###  THANDLE_INITIALIZE(T)
+```c
+ MOCKABLE_FUNCTION(, void, THANDLE_INITIALIZE(T), THANDLE(T) *, t1, THANDLE(T), t2 );
+ ```
+
+`THANDLE_ASSIGN` does t1=t2. `t1` is NOT a constructed handle. Example: it is a field in a struct that has just been `malloc`'d.
+
+**SRS_THANDLE_02_011: [** If `t1` is `NULL` then `THANDLE_INITIALIZE` shall return. **]**
+
+**SRS_THANDLE_02_012: [** `THANDLE_INITIALIZE` shall increment the reference count of `t2` and store it in `*t1`. **]**
+
+## THANDLE_TYPE_DEFINE(T)
+```
+#define THANDLE_TYPE_DEFINE(T)
+```
+
+`THANDLE_TYPE_DEFINE` introduces the implementation for the functions in `THANDLE_TYPE_DECLARE` (`THANDLE_DEC_REF`, `THANDLE_INC_REF`, `THANDLE_ASSIGN`, `THANDLE_INITIALIZE`) and two new memory management functions `THANDLE_MALLOC(T)` and `THANDLE_FREE(T)`.
+
+### THANDLE_MALLOC(T)
+```c
+static T* THANDLE_MALLOC(T)(void(*dispose)(T*))
+```
+
+`THANDLE_MALLOC` return a pointer to `T`. `dispose` is a function that the `THANDLE_DEC_REF` calls when the reference count reaches 0 in order to free the resources allocated by the user in `T`. `dispose` can be `NULL` in which case there are no user resources to be de-allocated.
+
+**SRS_THANDLE_02_013: [** `THANDLE_MALLOC` shall allocate memory. **]**
```

Comment:
>llocate memory [](start = 50, length = 14)

all of it :-)? Consider speccing sizeof(T).

Reply:
Ah I see, it probably needs sizeof(T) + the size of the dispose pointer.

---
In reply to: [385370942](https://github.com/Azure/azure-c-util/pull/16#discussion_r385370942) [](ancestors = 385370942)

Reply:
myeah, I could add "at least sizeof(T)" but that doesn't help much :(. I propose to leave it as is.

---
In reply to: [385371334](https://github.com/Azure/azure-c-util/pull/16#discussion_r385371334) [](ancestors = 385371334,385370942)